can't compile: driver/pulse_cnt.h: No such file or directory

### The problem

can't compile 2026.2.0, both frameworks

### Which version of ESPHome has the issue?

2026.2.0

### What type of installation are you using?

pip

### What platform are you using?

ESP32

### Component causing the issue

pulse_counter

### YAML Config

```yaml
substitutions:
   devicename: "gate"

esp32:
   board: esp32dev
   framework:
      type: esp-idf

packages:
   device_base: !include common/device_base.yaml
   wifi: !include common/wifi.yaml

status_led:
   pin:
      number: 2
      inverted: "false"

sensor:
 - platform: pulse_counter
   use_pcnt: false
   pin: 17
   unit_of_measurement: "W"
   name: "power"
   internal_filter: 77ms
   accuracy_decimals: 0
   update_interval: 180s
   filters:
    - multiply: 120
   total:
      name: "energy"
      state_class: total_increasing
      unit_of_measurement: "kWh"
      accuracy_decimals: 2
      filters:
       - multiply: 0.002
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2026.2.0
INFO Reading configuration gate.yaml...
WARNING GPIO2 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq/#why-am-i-getting-a-warning-about-strapping-pins
WARNING GPIO15 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq/#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Compiling app... Build path: /home/jirka/esphome/.esphome/build/gate
Processing gate (board: esp32dev; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50502.0 (5.5.2) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.2 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.10
Compiling .pioenvs/gate/src/esphome/components/pulse_counter/pulse_counter_sensor.cpp.o
Compiling .pioenvs/gate/src/esphome/core/component.cpp.o
Compiling .pioenvs/gate/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/gate/src/esphome/core/controller_registry.cpp.o
Compiling .pioenvs/gate/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/gate/src/esphome/core/gpio.cpp.o
Compiling .pioenvs/gate/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/gate/src/esphome/core/log.cpp.o
Compiling .pioenvs/gate/src/esphome/core/ring_buffer.cpp.o
Compiling .pioenvs/gate/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/gate/src/esphome/core/time.cpp.o
Compiling .pioenvs/gate/src/esphome/core/util.cpp.o
In file included from src/esphome/components/pulse_counter/pulse_counter_sensor.cpp:1:
src/esphome/components/pulse_counter/pulse_counter_sensor.h:12:10: fatal error: driver/pulse_cnt.h: No such file or directory
   12 | #include <driver/pulse_cnt.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/gate/src/esphome/components/pulse_counter/pulse_counter_sensor.cpp.o] Error 1
============================================================================================================== [FAILED] Took 5.20 seconds ==============================================================================================================
```

### Additional information

_No response_

--- a/esphome/components/pulse_counter/pulse_counter_sensor.h
+++ b/esphome/components/pulse_counter/pulse_counter_sensor.h
@@ -8,10 +8,10 @@
 
 #if defined(USE_ESP32)
 #include <soc/soc_caps.h>
-#ifdef SOC_PCNT_SUPPORTED
+#if defined(SOC_PCNT_SUPPORTED) && __has_include(<driver/pulse_cnt.h>)
 #include <driver/pulse_cnt.h>
 #define HAS_PCNT
-#endif  // SOC_PCNT_SUPPORTED
+#endif  // defined(SOC_PCNT_SUPPORTED) && __has_include(<driver/pulse_cnt.h>)
 #endif  // USE_ESP32
 
 namespace esphome {

--- a/tests/components/pulse_counter/test-no-pcnt.esp32-idf.yaml
+++ b/tests/components/pulse_counter/test-no-pcnt.esp32-idf.yaml
@@ -1,0 +1,10 @@
+sensor:
+  - platform: pulse_counter
+    name: Pulse Counter
+    pin: 4
+    use_pcnt: false
+    count_mode:
+      rising_edge: INCREMENT
+      falling_edge: DECREMENT
+    internal_filter: 13us
+    update_interval: 15s

